In [574]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sbn

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('max_colwidth', None)

artists_data_path = "../data/artists.jsonl"
sessions_data_path = "../data/sessions.jsonl"
track_storage_data_path = "../data/track_storage.jsonl"
tracks_data_path = "../data/tracks.jsonl"
users_data_path = "../data/users.jsonl"

artists_data_frame = pd.read_json(artists_data_path, lines=True)
sessions_data_frame = pd.read_json(sessions_data_path, lines=True)
track_storage_data_frame = pd.read_json(track_storage_data_path, lines=True)
tracks_data_frame = pd.read_json(tracks_data_path, lines=True)
users_data_frame = pd.read_json(users_data_path, lines=True)

Pominęliśmy analizę danych o przechowywaniu utworów, ponieważ nie wnoszą one żadnych informacji do analizy w naszym temacie projektu.

# Analiza artystów

In [575]:
artists_data_frame.head(n=5)

id             name  \
0  7b6Ui7JVaBDEfZB9k6nHL0  The Local Train   
1                      -1    Vishal Mishra   
2  6AETRCQep85mUtIlX7BaFV     Color Humano   
3  3PN8odvj48fnALdAsCKG3g           Crucis   
4  5SAeZRhgmJhgq196BR1Mna      Dúo Salteño   

                                                                      genres  
0  [desi pop, hindi indie, indian indie, indian rock, new delhi indie, sufi]  
1                                               [desi pop, modern bollywood]  
2                                            [argentine rock, rock nacional]  
3                                  [italian progressive rock, rock nacional]  
4                                     [folclore salteno, folklore argentino]

Atrybuty:
* id - identyfikator artysty
* name - nazwa artysty
* genres - gatunki muzyczne, które tworzy

In [576]:
artists_data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10657 entries, 0 to 10656
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      10657 non-null  object
 1   name    10657 non-null  object
 2   genres  10126 non-null  object
dtypes: object(3)
memory usage: 249.9+ KB


### Analiza id

In [577]:
# find artists with -1 id
artists_data_frame[artists_data_frame['id'] == -1]

id                                          name  \
1      -1                                 Vishal Mishra   
14     -1                                      Luz Mala   
25     -1                              Mono & Nikitaman   
61     -1                                      Rebelion   
99     -1                                           FSB   
107    -1                                       LouKass   
158    -1                                 Quatro por Um   
161    -1                                      Mc Sapao   
213    -1                                 Fred Pellerin   
229    -1                                          Gölä   
256    -1                                    Zalo Reyes   
257    -1                               La Pozze Latina   
260    -1                               Fiskales Ad-Hok   
270    -1                                   Dario Dario   
274    -1                                  José Capmany   
289    -1                                    Synkopy 61   
296    -1                                     Debustrol   
312    -1             Disney - Arielle die Meerjungfrau   
361    -1                        Fede Finn & Funny Boyz   
379    -1                                 Tommy Seebach   
399    -1                                Ramon Leonardo   
404    -1                                    Maria Diaz   
413    -1                               Sonia Silvestre   
441    -1                                  El Andariego   
448    -1                                 Kolumbus Kris   
471    -1                                   Luis Lucena   
472    -1                                        Haizea   
533    -1                                  Gordo Master   
537    -1                                  Las Carlotas   
539    -1                                      Els Pets   
549    -1                                 Musta Paraati   
653    -1                                Kotiteollisuus   
710    -1                               Viky Moscholiou   
714    -1                                  Panos Kiamos   
718    -1                             Miltos Pashalidis   
735    -1                               Christos Dantis   
746    -1                                  Fotis Siotas   
749    -1                             Anastasios Rammos   
752    -1                                   Yorgos Rous   
790    -1                                   Astro Bunny   
796    -1                                            鄭興   
826    -1                                     Ámokfutók   
850    -1                                       Dzsúdló   
907    -1                                     Superglad   
932    -1                      The Dublin City Ramblers   
942    -1                                           Ash   
955    -1                                    Mr. Bungle   
992    -1                                     Hayehudim   
1004   -1                                 Yardena Arazi   
1005   -1                                   Yoav Itzhak   
1006   -1                                 Gary Eckstein   
1022   -1                                  Lucille Crew   
1024   -1                                     Dan Toren   
1027   -1                                 Shlomi Shaban   
1065   -1                                Helgi Skúlason   
1124   -1                                Gunni og Felix   
1130   -1                                        Sveppi   
1132   -1                                        Grafík   
1155   -1                                   Jón Jónsson   
1163   -1                                         Attan   
1176   -1                          Modena City Ramblers   
1209   -1                                     Tricarico   
1231   -1                   Norihiko Hashida And Climax   
1233   -1                                          COCO   
1255   -1                                       はっぴいえんど   
1262   -1                                Stardust Revue   
1271   -1                                   Number Gir

Istnieje 557 artystów, których id jest ustawione na -1.
TODO co z tym zrobic?

In [578]:
print("unique ids: {}/{}".format(artists_data_frame['id'].nunique(), artists_data_frame['id'].count()))

unique ids: 10101/10657


Nieunikatowe ID wynikają z tego, że istnieje 557 artystów, których ID jest ustawione na -1.

### Analiza name


In [579]:
print("unique names: {}/{}".format(artists_data_frame['name'].nunique(), artists_data_frame['name'].count()))

# find artists with duplicated names
artists_data_frame[artists_data_frame['name'].duplicated(keep=False)]

unique names: 10643/10657


id               name  \
358    2xgbz6gqoIF4S6JZoRiH2y             Dalton   
390    38YroIcEOM2zsIEGdtmZYm              Bamse   
460    0Yk0PoyjQiEyIKgnaJFR0n            Traffic   
1093   3rUSk3kcpl4Nseu1DqFNpJ                 KK   
1384   3lPK4zqijDKAvZkdlmZRfN              Mango   
2076   4ppDL5W65K1HG6EF7yaDLd  Little River Band   
2352   53Thxvlr6imD5y3lxouOu4            Phoenix   
2546   363VbwpX9anvrThJ2qDwBy                TNT   
2721   1CqOLQmjzVWXQTiIN5Wucs                TNT   
2757   7c5278WShlxTk3vS9XoTf8              Wings   
3967                       -1               Eloy   
4314   1FizW0swe8ODtXvEOHXbHX                LFO   
5192   3Vszg64mczF1dWEcW2dU5W            Luciano   
6283   2jFXKAH9A9H8lVnJkkLxXl              Omega   
6495                       -1            Traffic   
6803   0eet2VBconGGtNYJ0CBgHu               Eloy   
6959   0UCa56b7hiX05ZWC5BH2zo             Nikita   
7561   1UjxAZqzphB1tsMb1aWBj0              Omega   
7887   1xU878Z1QtBldR7ru9owdU            Phoenix   
8398   3sFhA6G1N0gG1pszb6kk1m              Wings   
9113   1V95p8oWGxXlpcFRYqXV3w              Bamse   
9262   6clbbhnIqpHnqxwtOWcilg  Little River Band   
9286   3CJKkU0XuElRT1z8rEtIYg            Luciano   
9455   4fEkbug6kZzzJ8eYX6Kbbp                 KK   
9461   5iBphIrmXEjfrqiuz5y6pY              Mango   
9869   2M0T4a1pkOC5nifN9W6e9e                LFO   
9944   7tS2WuFuYfT8wcw0dy63D9             Nikita   
10467  25JJnYKQwlu4uPHJP0pN5l             Dalton   

                                                                                                                                                                                                                                  genres  
358                                                                                                                                                                                                                 [classic danish pop]  
390                                                                                                                                                                                                              [barnmusik, bornesange]  
460                                                                                                                                                                                                        [estonian pop, estonian rock]  
1093                                                                                                                                                                                              [classic icelandic pop, icelandic pop]  
1384                                                                                                                                                                                                                    [lithuanian pop]  
2076                                                                                                                                                         [album rock, classic rock, country rock, folk rock, mellow gold, soft rock]  
2352                                                                                                                                                                                                                     [romanian rock]  
2546                                                                                                                                                                                                             [glam metal, hard rock]  
2721                                                                                                                                                                                                     [euphoric hardstyle, hardstyle]  
2757                                                                                                                                                                    

Istnieje 28 artystów, których nazwa jest zduplikowana.
TODO co z tym zrobic?


### Analiza genres


In [580]:
# find artists with null genres
artists_data_frame[artists_data_frame['genres'].isnull()]

id  \
24     058bioyBUzSkKCfrqvODLu   
26     3N4ZBIXZfvF0VLgEclgJvX   
85     3fxtMTmbY0jgtjHDuXspus   
122    0cDZ1YAlkkIn0Z7BjiEUj1   
125    5CorrQPbOid8RKf5rCpxG7   
135    2u7kjqj92WlpCSAKiOeZ6I   
169    28rjo4eRvpmLW0EsCIy6FW   
187    4fk1ToCb0m7irMHCaZhIrM   
190    0hGCbqEUzKxBUuJ7Q8ZFvK   
208    3IYryWetzBVax8oJPkgXMk   
217    5x0wy0trcqYLvwUG5PFmQL   
231    7wkPBPwF9oOZJ8lEbQjIVt   
249    7C7z1OMQ7TOtFBkwzK0zX9   
254    0oGbFTtR8UqMpgDFLpFKc1   
357    4dj3Rt7naflzJgPcFkOJH5   
365    2Fgt9tP2vJusWCaJK3e9QR   
373    1SrmGTb6i2OfOmNkbX3Ft4   
380    44os6d6JylvrXEyQLYeqGr   
417    46aJHc5ahjcNZfiS44i1zo   
487    1r5SGuosYsz2ruhoBgJ4iF   
490    7nOYYLDwTAvcZ4Hy5tOry0   
537                        -1   
574    627dKDcsgfO25s2ScC4sgQ   
613    6JlgvkzzxFaMfDpu7zxUav   
614    7zD8bBoHnqnQsjYWEOe3V8   
646    3hMNlyrxz0vtDz70ewLHgl   
654    2ECoOJ3zAmnbZoCwTaf9A2   
676    5Jh4UFOCWwleD3kcd88xyh   
712    1crGTvNXrAffvbpXZyOCKx   
745    5HRyAD6yCLqZlzh4R4ds1O   
772    4C7zRUPI7U5RVXGbpgeg2b   
798    6pnkoq165sVlU8H3WZKnhn   
807    3eeg2cVDU1rRPrvIwa0FlX   
837    68yBeY4QowunDr8foW1VaI   
839    32E92dXwYIiCE5ORdXRVOD   
871    77X41f0zoHXGhC3a3UvdJk   
874    4RLzLs2SHeyA0nLLlAJBg4   
884    126aQ4k98tuE3BXgfPwHpX   
896    5sFfUahhUdAKlrr2AUyMx6   
929    5w2TwcoIzksdooTyczidj3   
941    6lyku6lhbB3tLc3Dghs5CM   
945    4zPGlyitRvJo7iz6OKlETa   
953    6a2HcrBBLrhSq57wzuoXYr   
964    1F102kNzMqsmOpF7AfFmm5   
976    5ETZ4zl1tAJe8zCTDEaoGO   
978    110wHDAsaXDaKKdUztwGlE   
991    3giniJpo19Md4V5Plataq2   
1018   4VxlxOXkJFI9QRdBHUDIC4   
1038   2Eij3bHbG52JBSEqXSwga3   
1046   6eRpJMvXvYgm6apZRfoQWk   
1053   2q1LRGJHpFxovU8Tz6OgRn   
1094   5IRON8eFfbctCsLo6HlSdQ   
1095   1rJkz5vopfGxTUGFNB3o4G   
1097   01w8eDOz2L9SFgv1ukZPRv   
1099   0my8yB6MM8YVbIR8A96GY3   
1161   3vvugAMkG44OwAjfds1QpR   
1165   5XcM6fOuamexiFlGtmMRwq   
1187   3uQsgKw2TZkZRbxAeEdZfW   
1258   2E3u3DFA1rwICiHqlXJ05L   
1352   5ctTSlA3JZKProh0uSzTa8   
1357   5fd9qkcJWbNiFtlGGaoVIr   
1432   4oopAFtLdKXCoq2XoEjDZD   
1437   1KiCGFfdGc5RDccoRXHFjw   
1456   7GIvsDVOV1bjghuisC9opx   
1469   5A9NWhs9ydYZPGgvcWI8Ms   
1471   2ow8nzGoMwOtNtiupXpANu   
1474   5sWCGfIxmNiMAl6n8kdbkb   
1494   1QRod4LR53Ouoq8XoGdF4Q   
1504   1VeH892FTsVyB6XAnMvEbL   
1508   22XVSTMZerjq8l2QxXNMFY   
1546   2T6hbG0qQTyZglPgH7UEMP   
1556   3FsNh5N2AxOGA1UU5IAnlG   
1592   0bYI0rlhxJWiOD2ZRKbb7w   
1602   1F30RWETubxDmbyfxeVPg7   
1626   0ob2mO2kOzEdaGLYSth531   
1663   4WLGcWrkSExCqILxDk7ol6   
1686   2lzeV1nxm8MB7z85L53JV2   
1709   5icZHFQjI6SRsvXNOKUXZ1   
1734   6Y99TxJKt19wJFbmpoRzLp   
1763   1QfTCSQTI4rrK027UaiSQ6   
1786   7s5xlRV8yGHwxkPggVKdN6   
1793   5gsErkGsLeCYCSK0Pxj01N   
1822   1c9LVBvYq58ViD8qdLml7A   
1838   1dmEUAQBopYUrsMnQnScWL   
1857   72eP0W3rIhkxd0NHGg4w4u   
1873   58BqVfV4jCaxcbdZnmUvxK   
1922   6ELAHXutcbgmM3v0Bg2nei   
1947   0Nf2qv7lBcOJ2u3tH0YPXX   
1998   0Utyuwr2dSRI7b7JDaBfbK   
2017   5pbzj8JGUpUCMHESd1yHFv   
2036   56CaueCWXmgGQ9sFaUTeMa   
2090   6wYJgFEDVZmiCywTxOmoEe   
2110   7n1EM42Fseebd9H2p8eWWS   
2114   1RSKoT6hnzaLjvL0LCF0ew   
2133   0FRKTwQSToXpCxYMhyUzYY   
2135   3O9bFJBTRddC4HOsZixhuw   
2179   28WXX79U4jzq1kfjx5QXPy   
2182   1nEmAvnOKMBYCziZQNb5B4   
2191   3kUC3a0wTkjsdkJiytULkN   
2202   2HVjYynnJqLIg4HVV2A31m   
2215   2JDzHpbTcm5cu8Ne2SDYo4   
2224   4XI9RntQXI5Ly7j87g9y9I   
2256   1ts2oIXKCpWSRFnR78bulp   
2277   4Scx8KcZlLA5ORsj5Gejci   
2280   5O9SEPgz10bQCeiIctR08i   
2319   5uOMOTzmfhliUjnyiJh0kn   
2356   6XuwlE6QNkhGKauGl4jM7t   
2357                       -1   
2367   5W6FFILqUpVfpsZC9KadWN   
2372   1Tm5STqbzJ39okZK1y6UWK   
2379   6l6cvAPpAXHxNHYzSHHTCu   
2404   0PqNs9XCK5D16ihtRMuOoU   
2419   7LMlaqzutxVEEEung8DPbE   
2434   2iHiqwlS40GTFt8eELeoHr   
2444   2OU0OH164ypvP4G3AYynXs   
2447   4DzKphc3l03DORh036Vhrw   
2458   0IyeWjNLB0Lxw0MgCEZtHw   
2474   2GqX0GnrKdi6Z2X2OReYOK   
2481   4w8YQxx0XQfWZgQwRSqcoE   
2487   7zYj9S9SdIunYCfSm7vzAR   
2507   1bufaOa1xsdvxGxeqQJtzW   
2

Istnieje 531 artystów, których gatunki nie zostały określone.
TODO co z tym zrobic?


# Analiza sesji

In [581]:
sessions_data_frame.head(n=5)

timestamp  user_id                track_id event_type  \
0 2023-01-04 11:46:25.000    101.0  78ybROgQ2V3TySb8ZfPNdg       PLAY   
1 2023-01-04 11:49:32.533      NaN  2lkoxPYi7o9OlaGP9bVI09       PLAY   
2 2023-01-06 04:38:08.000    101.0  0juIv1JPqoCrWku2nIfMiA       PLAY   
3 2023-01-06 04:40:49.544    101.0  0juIv1JPqoCrWku2nIfMiA       LIKE   
4 2023-01-06 04:44:14.267    101.0  4srmb7zgB7eKWLUL96xftd       PLAY   

   session_id  
0         124  
1         124  
2         125  
3         125  
4         125

Atrybuty:
* timestamp - czas rozpoczęcia sesji
* user_id - identyfikator użytkownika
* track_id - identyfikator utworu
* event_type - typ zdarzenia (PLAY, LIKE, SKIP)
* session_id - identyfikator sesji


In [582]:
sessions_data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3971 entries, 0 to 3970
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   timestamp   3971 non-null   datetime64[ns]
 1   user_id     3776 non-null   float64       
 2   track_id    3776 non-null   object        
 3   event_type  3763 non-null   object        
 4   session_id  3971 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 155.2+ KB


### Analiza timestamp

In [583]:
sessions_data_frame['timestamp'].describe(datetime_is_numeric=True)

count                             3971
mean     2023-02-09 22:20:59.431924736
min                2023-01-01 01:12:53
25%      2023-01-22 12:22:54.168499968
50%      2023-02-08 12:59:38.910000128
75%      2023-02-28 06:16:43.153500160
max         2023-03-24 06:28:07.464000
Name: timestamp, dtype: object

Sesje rozpoczęły się 2023-01-01 01:12:53 i zakończyły 2023-03-24 06:28:08.464000.


In [584]:
# find sessions with timestamp, but without user_id, track_id or event_type
sessions_data_frame[sessions_data_frame['timestamp'].notnull() & (sessions_data_frame['user_id'].isnull() | sessions_data_frame['track_id'].isnull() | sessions_data_frame['event_type'].isnull())]

timestamp  user_id                track_id     event_type  \
1    2023-01-04 11:49:32.533      NaN  2lkoxPYi7o9OlaGP9bVI09           PLAY   
9    2023-01-06 04:47:01.091    101.0  7fcD9YsAiBbtVyKe9Tqll6           None   
10   2023-01-06 04:47:17.285    101.0  7fcD9YsAiBbtVyKe9Tqll6           None   
15   2023-01-06 04:57:05.043      NaN  0Um0731HrqdBsyGs3WdmJX           PLAY   
32   2023-02-08 12:57:06.312    101.0                    None           SKIP   
52   2023-02-14 16:51:43.190      NaN  0Lh0VkvZ3LJPR5fUs0pigo           LIKE   
59   2023-02-22 17:35:32.492    101.0                    None           LIKE   
70   2023-03-07 16:11:35.000    101.0                    None           PLAY   
71   2023-03-10 22:02:57.000    101.0  20IIOqNWjz9UL3EyxDgMfK           None   
75   2023-03-10 22:05:46.984      NaN  2KqeHIAjaeeeUUJMzlv698           PLAY   
80   2023-03-10 22:14:54.532    101.0  0IFbyuvJUXroM8UkZ2Pnsk           None   
83   2023-03-13 16:00:04.000    101.0  3XOJhbofb4qTZdt0RLLrUq           None   
85   2023-03-22 00:33:50.000    101.0  7iQVotSlT0XFgcth4K1O5Y           None   
89   2023-01-14 05:13:34.480      NaN  2u8OFtCDsutFApNvEEx7b5           SKIP   
92   2023-01-14 05:17:14.063    102.0  4O5GerAkdrXZCCGdSK654u           None   
107  2023-01-31 04:38:31.947      NaN  6Y09N5LKkGxeVi0mfEmA7q           PLAY   
115  2023-03-14 11:14:49.000      NaN  2jHlixeBq1ReAxh22WG8KO           PLAY   
116  2023-03-14 11:15:14.419    102.0  2jHlixeBq1ReAxh22WG8KO           None   
121  2023-03-14 11:26:30.668      NaN  2pxqnc6MveoeOQSlX7yWkh           PLAY   
141  2023-01-30 16:30:49.976    103.0  34udgpV3Rdur2RNT3VSYTf           None   
143  2023-01-30 16:31:12.169      NaN  51RmAIH9LqZeZIf76v0N4g           None   
146  2023-01-30 16:34:18.068      NaN                    None           PLAY   
153  2023-01-30 16:45:34.116    103.0  6bkOMcBuSjl2cftYa2DrFj           None   
158  2023-02-06 07:06:43.989      NaN  2nTlUSqBKN7FVsqD9y4MSf           PLAY   
160  2023-02-06 07:10:15.669    103.0  00Amd2EiGo17YoZyOjk3VV           None   
164  2023-02-06 07:19:44.228      NaN  2VGfcdXmSNe2jjfVISIPCR           PLAY   
165  2023-02-06 07:20:15.330    103.0                    None           LIKE   
173  2023-02-14 03:28:00.514    103.0  0obhPhoZZpTRTt8uRQAIwE           None   
186  2023-02-14 03:56:56.180    103.0                    None           PLAY   
200  2023-02-14 13:04:36.391    103.0                    None           PLAY   
201  2023-02-14 13:05:15.827    103.0                    None           LIKE   
205  2023-02-18 21:34:38.368    103.0                    None           SKIP   
208  2023-02-18 21:44:28.428      NaN  4QRsnJRKgHF6gRPZQzNMu8           PLAY   
212  2023-03-04 10:42:26.616    103.0                    None           PLAY   
230  2023-03-13 22:44:14.145    103.0                    None           PLAY   
232  2023-03-13 22:53:00.135    103.0  1PpJ3v90IgcADXsB3VIq9F           None   
237  2023-01-08 22:17:47.893      NaN                            BUY_PREMIUM   
247  2023-01-10 05:29:24.864    104.0  4SKXlJ5NN37YwJOHa1Dz7a           None   
256  2023-01-10 05:44:45.555      NaN  37pmVaDTYUIxp6XwVQOy5C           PLAY   
263  2023-01-12 07:53:52.263      NaN  2HDdFE3LzgZ2ZWVXS0jEQF           PLAY   
264  2023-01-12 07:57:49.356      NaN  28qvZWA2nwiGKjCS3GaRxo           PLAY   
268  2023-01-17 19:48:52.000    104.0  4heoK6zUM7ElZyytFgnrrk           None   
279  2023-01-17 20:05:34.733    104.0  6dFvshsBNh4XDJXSWwRNel           None   
298  2023-02-16 09:02:22.575    104.0  6eZMwVg1LkthqrOwliEm1z           None   
302  2023-02-16 09:06:03.903    104.0                    None           SKIP   
304  2023-01-12 14:04:53.000    105.0  1bXVWeaDqCi1knnWCOdCWH           None   
307  2023-01-12 14:05:37.345    105.0                    None    BUY_PREMIUM   
312  2023-01-12 14:13:03.756    105.0                    None           LIKE   
313  2023-01-12 14:13:39.626      NaN  56BTcff0EBl5oW0y1VY0q8           SKIP   
314  2023-01-12 14:

### Analiza user_id

In [585]:
sessions_data_frame['user_id'].value_counts().describe()

count     50.0000
mean      75.5200
std       30.6436
min       20.0000
25%       50.0000
50%       75.0000
75%       95.7500
max      148.0000
Name: user_id, dtype: float64

Średnia liczba sesji na użytkownika wynosi 75.5, a maksymalna 148.


### Analiza track_id



In [586]:
# TODO


### Analiza event_type

In [587]:
sessions_data_frame['event_type'].value_counts()

PLAY             2242
SKIP              825
LIKE              619
BUY_PREMIUM        47
ADVERTISEMENT      30
Name: event_type, dtype: int64

### Analiza session_id

In [588]:
sessions_data_frame['session_id'].value_counts().describe()

count    636.000000
mean       6.243711
std        5.611185
min        1.000000
25%        2.000000
50%        4.000000
75%        9.000000
max       33.000000
Name: session_id, dtype: float64

Średnia liczba zdarzeń na sesję wynosi 6.2


# Analiza danych o utworach


In [589]:
tracks_data_frame.head(n=5)


id                                     name  \
0  4LdczeQKU4ojwFJd6QWGnW  Taking Care Of Business - 2005 Remaster   
1  7F20EOraDwU8KvNcyzUKVz                                     Leef   
2  5EMGRFIidh98v9eXj6QnOU                              Dime Que No   
3  5XFfcpc0V1fqHeiEjJd4y5             I Hate the Capitalist System   
4  4fsmo5mukKd4vafcHN41KP                       Hai la sârbă roată   

   popularity  duration_ms  explicit               id_artist release_date  \
0         1.0       177853         0  4AgFg0hHythyklEs4t7XKt   2011-01-01   
1        42.0       217903         0  5l2Xy4aUoJDRSpsYHyOumD   2015-06-26   
2         NaN       266013         0  0h1zs4CTlU9D2QtgPxptUD   2012-01-23   
3        28.0       196387         0  7htyUKqAPLpbTA9nkEhGRf   1973-01-01   
4        24.0       223013         0  4a0FNCbvDEoeHYFfc045W4   2008-01-01   

   danceability  energy  key  loudness  speechiness  acousticness  \
0         0.554   0.847    7    -7.237       0.0633       0.00789   
1         0.673   0.727    9    -6.587       0.0401       0.28200   
2         0.535   0.648    0    -5.842       0.0364       0.36900   
3         0.454   0.115    5   -15.617       0.0405       0.79700   
4         0.509   0.699   10    -4.935       0.0372       0.68600   

   instrumentalness  liveness  valence    tempo  
0          0.000002     0.183    0.733   90.655  
1          0.000000     0.178    0.782  121.016  
2          0.000000     0.075    0.311  142.486  
3          0.000000     0.103    0.366   86.761  
4          0.000016     0.267    0.949  174.827

Atrybuty:
* id - identyfikator utworu
* name - nazwa utworu
* popularity - popularność utworu
* duration_ms - długość utworu w milisekundach
* explicit - czy utwór zawiera treści nieodpowiednie dla dzieci
* id_artist - identyfikator artysty
* release_date - data wydania utworu
* danceability - współczynnik "taneczności" utworu
* energy - współczynnik "energii" utworu
* key - główny ton utworu
* loudness - głośność utworu
* speechiness - współczynnik "mowy" utworu
* acousticness - współczynnik "akustyczności" utworu
* instrumentalness - współczynnik "instrumentalności" utworu
* liveness - współczynnik "żywotności" utworu
* valence - współczynnik "pozytywności" utworu
* tempo - tempo utworu

In [590]:
tracks_data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21608 entries, 0 to 21607
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                20496 non-null  object 
 1   name              20557 non-null  object 
 2   popularity        20505 non-null  float64
 3   duration_ms       21608 non-null  int64  
 4   explicit          21608 non-null  int64  
 5   id_artist         20526 non-null  object 
 6   release_date      21608 non-null  object 
 7   danceability      21608 non-null  float64
 8   energy            21608 non-null  float64
 9   key               21608 non-null  int64  
 10  loudness          21608 non-null  float64
 11  speechiness       21608 non-null  float64
 12  acousticness      21608 non-null  float64
 13  instrumentalness  21608 non-null  float64
 14  liveness          21608 non-null  float64
 15  valence           21608 non-null  float64
 16  tempo             21608 non-null  float6

### Analiza id
TODO

In [591]:
# find tracks with null id
tracks_data_frame[tracks_data_frame['id'].isnull()]

id  \
7      None   
23     None   
83     None   
104    None   
156    None   
161    None   
188    None   
191    None   
214    None   
227    None   
228    None   
231    None   
245    None   
258    None   
274    None   
287    None   
337    None   
344    None   
348    None   
351    None   
375    None   
402    None   
407    None   
414    None   
452    None   
456    None   
503    None   
504    None   
507    None   
513    None   
519    None   
569    None   
597    None   
616    None   
631    None   
722    None   
725    None   
752    None   
764    None   
768    None   
781    None   
797    None   
823    None   
850    None   
860    None   
892    None   
896    None   
910    None   
919    None   
925    None   
984    None   
985    None   
1004   None   
1019   None   
1046   None   
1107   None   
1126   None   
1135   None   
1139   None   
1182   None   
1208   None   
1218   None   
1284   None   
1328   None   
1329   None   
1352   None   
1354   None   
1373   None   
1385   None   
1424   None   
1434   None   
1461   None   
1470   None   
1478   None   
1507   None   
1544   None   
1555   None   
1577   None   
1590   None   
1626   None   
1640   None   
1697   None   
1732   None   
1743   None   
1751   None   
1772   None   
1814   None   
1817   None   
1852   None   
1873   None   
1883   None   
1894   None   
1951   None   
1983   None   
1996   None   
2023   None   
2045   None   
2053   None   
2070   None   
2107   None   
2113   None   
2116   None   
2133   None   
2158   None   
2187   None   
2192   None   
2200   None   
2219   None   
2230   None   
2256   None   
2257   None   
2283   None   
2312   None   
2463   None   
2471   None   
2482   None   
2487   None   
2490   None   
2498   None   
2579   None   
2617   None   
2629   None   
2650   None   
2655   None   
2683   None   
2689   None   
2708   None   
2731   None   
2742   None   
2746   None   
2749   None   
2790   None   
2794   None   
2806   None   
2828   None   
2838   None   
2842   None   
2857   None   
2878   None   
2890   None   
2892   None   
2895   None   
2901   None   
2918   None   
2924   None   
2942   None   
2987   None   
3001   None   
3010   None   
3018   None   
3019   None   
3030   None   
3078   None   
3120   None   
3122   None   
3159   None   
3167   None   
3179   None   
3192   None   
3194   None   
3196   None   
3241   None   
3252   None   
3254   None   
3279   None   
3292   None   
3310   None   
3323   None   
3393   None   
3399   None   
3443   None   
3474   None   
3476   None   
3481   None   
3493   None   
3502   None   
3521   None   
3526   None   
3533   None   
3537   None   
3539   None   
3544   None   
3545   None   
3546   None   
3549   None   
3565   None   
3566   None   
3588   None   
3601   None   
3625   None   
3640   None   
3641   None   
3660   None   
3662   None   
3664   None   
3672   None   
3738   None   
3739   None   
3873   None   
3913   None   
3974   None   
3986   None   
4004   None   
4027   None   
4034   None   
4040   None   
4098   None   
4100   None   
4109   None   
4113   None   
4127   None   
4129   None   
4142   None   
4143   None   
4146   None   
4147   None   
4166   None   
4174   None   
4200   None   
4202   None   
4206   None   
4227   None   
4265   None   
4290   None   
4299   None   
4315   None   
4347   None   
4351   None   
4355   None   
4385   None   
4409   None   
4462   None   
4495   None   
4501   None   
4548   None   
4552   None   
4559   None   
4602   None   
4608   None   
4633   None   
4634   None   
4659   None   
4678   None   
4733   None   
4746   None   
4763   None   
4772   None   
4776   None   
4780   None   
4784   None   
4786   None   
4792   None   
4865   None   
4873   None   
4894   None   
4899   None   
4911   None   
4921   None   
4941   None   
5103   None   
5107   None   
5115   None   
5131   None   
5140   None   
5159   None   
5161   None   
5209

### Analiza name
TODO

In [592]:
# find tracks with null name
tracks_data_frame[tracks_data_frame['name'].isnull()]

id  name  popularity  duration_ms  explicit  \
114    7wolKJcoDqcyCIC9sLy2l3  None         8.0       275737         0   
158    4Y4XzfNAzOqax0YcXxkP0u  None        33.0       280693         0   
174    3yV7OpqJxlFd7yWnOEDOYU  None        13.0       207067         0   
191                      None  None        40.0       213160         0   
200    3FSaOL1i67DehIF1gqKkQv  None         2.0       178027         0   
205    6oPqaiLKPdQj2ecBpMgM07  None        35.0       222733         0   
210    072nOX5nDAGalF8TC30EFv  None        30.0       190400         0   
216    4PCvrovUYITtRCqV2vv7pu  None        36.0       236520         0   
226    4u5MR81TPxZ9odymoLy8bb  None         0.0       165897         0   
237    0zoNNZmhsJXVBJEXeTg9vJ  None        16.0       225693         0   
268    0ULQ4oD3hGc5574sWfdgdK  None        35.0       217627         0   
283    08c3wmdjKSjxTJY9cHSts7  None        56.0       167907         0   
327    6U7q3t3ncfH7fKkhizUXOi  None        22.0       217333         0   
344                      None  None        38.0       451427         0   
359    3mFn0joMJjQyKcNwLXDPGA  None        33.0       226844         0   
391    6DlKoOpbwNyZ2QBqQ8HK12  None        28.0       327173         0   
393    4ti5pZ0zs1PDf1eQswskFB  None        49.0       277040         0   
430    21q9esRXK0doM41hAHXgkv  None         3.0       155640         0   
443    6bkvIebUYYgPo7I3t4eoQf  None        39.0       239267         0   
487    5hrPTzYAdfnO0EgjCO7uTQ  None        38.0       194467         0   
497    7ye7WipkXmLSCNcGms92qE  None        21.0       150827         0   
523    5DZk3APO6CwH0pFmol7Y1A  None        24.0       311693         0   
526    35hbQXI15TrRQW65wYJqc1  None         0.0       143107         0   
576    0gssidC4OIzAqNQE1Ne2CW  None        37.0        62396         0   
608    6V6p3kX685TlL51aeSgzpa  None        40.0       144680         0   
610    7gPjFi5znMxv8hfztiKCHp  None        23.0       186467         0   
631                      None  None        51.0       157040         0   
632    41mQO9fXQFBNjVjtp8DJsj  None         4.0       219880         0   
640    2oW8pBaoHKnG2zf7auryUH  None         NaN       156736         1   
696    0hgXyfd8dzmVxQ4r81lZOc  None        36.0       292467         0   
702    0lkw6BzFFLMmve7LVl7oWz  None         0.0       148953         0   
711    7gU3JGVmZa53CP3GO6FH7m  None         0.0       134507         0   
725                      None  None        27.0        77177         0   
746    3SMlhUdNu5m8yfvNUBWv8c  None        14.0       587733         0   
792    7s7LzVu7S3OQUWXmzyWi5c  None        27.0       125893         0   
814    3FdyPhltrJDyoZ4iz31ucs  None        38.0       282682         0   
826    7fW1OwQPexk5OWCj7CSpqU  None         NaN       125400         0   
840    0AAsZvbbE1dGnveap3LJnz  None        16.0       150867         0   
843    5APXeZDvjbbraoRXOqPDPN  None        11.0       298440         0   
872    3ISNvUn2gTemicNE15ccfT  None        44.0       216373         0   
882    0iOQWMVvNGFCWTpcXNxFyQ  None        43.0       191347         0   
893    0dVtyfsWmRRW1B2M4lox2H  None         9.0       209627         0   
924    6xBb8VL1NIj6pq01KXwiAb  None         NaN       149082         1   
999    7hpa3Bcjt8Or7tC7TtOIcs  None        29.0       183867         0   
1014   6OPNvqvtQf9QT5zPh8wDk7  None        53.0       219333         0   
1042   3zcpi0blxOLxMy2emwgiRM  None        40.0       228692         0   
1051   2KXOfp7tKmkiO71awDRBtT  None        34.0        92867         0   
1057   6pYRMlsqZypESVEl9y68gP  None        26.0       230040         0   
1068   0tbjiOUl4k492KPdWZS9sy  None        69.0       210123         0   
1072   3quIECsOEVvn4qwo25on5x  None        37.0       232467         0   
1087   3YVqowQcP5PysoI7YmK5qF  None        21.0       311600         0   
1097   2uKpZOUTtGLcWhFkGz1YbX  None        32.0       180584         0   
1114   0QepvU0N2fC2B5uIPafO1q  None        58.0       264040         1   
1144   3dCg1F8dLVNnNEpMwaEN8d  

### Analiza popularity
TODO

In [593]:
# find tracks with null popularity
tracks_data_frame[tracks_data_frame['popularity'].isnull()]

id  \
2      5EMGRFIidh98v9eXj6QnOU   
12     0CJQzybtZV8qql6Z1YtdLg   
21     1SsgCimc43JJfajCXSGfwY   
115    0w5HJfvdcyw5aYJTyq6TJ8   
121    6BnQFRnHUkh9BU70bx59yG   
175    76c2rWXt6MlW8lUlHufJ4f   
193    2BLI4Z9k2GU7HF7Oj4Schg   
219    4wendPqEPX9hknmLYLzWp4   
227                      None   
247    0gWqlhTvn7XbkseUhKA3bl   
274                      None   
284    4tUvalPiIosJVC3iiNICzD   
311    5dwc5GPn3IZMnfEa9o5LsV   
313    7aExqfgu7jhqPIxdeQL2PE   
324    6oEwV0l34zhnC7OmAqmHjY   
330    0mGBA5vtFH58DucfqaW6oG   
342    3j3Zrq7CiZtidBdcDrVIrd   
348                      None   
374    73M2Vb5MfZh8iGKudkMtlw   
395    1jiSO7mpIQ8cb4Qv0uloVG   
440    3V33ldRN2gRQyzojTwjlWP   
485    73RjLOFqp2qbJitNbAbJVh   
486    1NgRBQQD8ZHy6lv79KyeZO   
506    5YCQGlR3QNIbtX75DW5mTx   
509    6BBzqZUektI7Z1BQXdJFIr   
510    4UK1g9Z1lWedkmIQuSyCe5   
537    1JkijYB4a2jrm0YlshWwwv   
545    5cmMh3PpN0A2LjAQIzBaoy   
571    2imSVUDEzlZmIRVVuT8zFQ   
578    5Sda0dVPJ10rjvw9zJnpeX   
591    6G1xl1e4a33zPYiYKrGIWm   
601    1ptU2y2UTWMZvK8W4KDPua   
640    2oW8pBaoHKnG2zf7auryUH   
678    1QOaJtgquGNaI6heywpdTW   
710    1aBrTPZmwUcGWX35lxJWFt   
740    7BCSlcMVc7OOoTCXw8JzPW   
760    1DtIQXGc2xt0COxPW8Jhrs   
795    46dgEyMQRTp60Kvw5k5yLt   
826    7fW1OwQPexk5OWCj7CSpqU   
846    3osoedzMlnwf90Qu0T11rx   
853    4qjwP8r20YF4UilTAiTJua   
869    6pGb377TQ02u68C39daWjU   
924    6xBb8VL1NIj6pq01KXwiAb   
933    5ofSK371EJa2gxYkru8oAd   
981    2JI7FsOU0FpzK7HqM6Ton6   
988    5emy6FZPDS3b9XNUlxISBi   
1018   2HmR8n6D6yaMlTMRb8BOFe   
1060   2cPZvwW2IR913A8sSDil84   
1062   3UHCozHfCVBFtSlKV5p0zh   
1063   2lh0pDr0nxMzAXISb0eycX   
1071   3xrEmo0j6faj2PpPrpqOrO   
1079   2RhRC9oMxEm5k0EJCmokSf   
1089   3xsop8PrXeM1LHYn7d2Avy   
1093   2MmLlygV0PTD8rnsgtpbYy   
1095   1Q7YzsHwwrTH1V3SPNjqmk   
1098   0UvOvD6ulWO1Lx33eAUCMI   
1126                     None   
1192   37vwjNlC6zsNuD0dswKgKX   
1196   1BAfKSDGqKLjm5ExkNq0FS   
1202   7Ln0OEpEiFDMwxHfEMP5aM   
1205   6h5oWp72pBSmfni13edX7k   
1232   1EsvnviFydQYUXisXjZINy   
1296   2Sjs9Kb2nKqbDat14bI6al   
1302   7hz1CZ9EqizK9QjAIQFqRi   
1325   7MAn3HImQBh5FQoENmlJ6Q   
1337   21HfUACOv1lgBiYocruoW8   
1361   59t9MAGzChjiD6Eowgemit   
1379   6oWUM4OHFV1F6tft35yzPt   
1389   5Qe7NHxeLAn8KoLTNLSdwe   
1433   1V5sk2uL3kw1uzpLQOLZcj   
1435   5XdczKVzK5UrLVHoMRH7R9   
1460   2s0XwRidJ0uVeXThGbeCCk   
1482   69np0vEH8DIRlKZqSfnCBi   
1501   4KJ6V1umnYtXASC7poAjwa   
1510   7HHhnM3x7evt4cq6ZBq6BF   
1545   2rPWQKW9q0mmi10LJHQ7vI   
1570   1EAb9MI3olU3eiyGhrMylw   
1579   67VWKSQv5P2bOlD6aqzHRM   
1587   44hgnwIPlGiXhS2aYoBPTB   
1590                     None   
1622   1GHXn1cxSMhFBI823DQoSe   
1638   1ueutfc6deduXZwqSu9qlw   
1644   69ZlHOU7MoMVD0D7WxJkxe   
1653   4fQHH8HlNoPBxRi5vorTIf   
1678   5c6DkbuXXka6WBtlR2oaGd   
1690   6anOXnPqtyBLtznxJphtwW   
1701   0Rog1z4YLdTtPcBAizceHQ   
1723   2ZfweMhYT4YO5xJHo2cXLK   
1738   2N24vhb3CGNcMdNuOKmAqf   
1793   5Pq42593Ef6UpMNaceIm7i   
1795   5wJSqlciE65IfOUW5DlqYg   
1804   2ZQBueVhI5teuS0MppBE63   
1827   7coxV57ERBKImh8Yy383IT   
1857   06HfXkX9LYRXcqhVWqYBra   
1888   0g06Zr1lI0c3LqvL3Q5WCV   
1923   17PdojRsFJCQcEdg8gURhS   
1935   2K2XvwoIKY8zP8mCFuw8qa   
1966   2g1EqSTNtd6Q9qA9vrH08d   
1994   4NkB3G5RGC77hbDoeAvSGH   
2001   0v2YFA6OlpbB3BBAvzbGRS   
2016   1DK0aJux0IXWBPCBP4NTal   
2050   1nF7wnmwOCu8y8KWIZNYvg   
2063   6aKzi3zSmZ6Lo7WNf4cFF0   
2075   0nVD7IBnTnA8pMoTjHrMgA   
2087   4gh78osufvCssjiZveH41y   
2135   4jSq795vEU1RM6AzHicKCU   
2138   5RsdPau1gaKqe8ZNRFcVvT   
2158                     None   
2171   3Omc1JskWu7w1bJqBcB5dq   
2176   5EsbTLcWh4nE5hg34eLqz9   
2178   6EHIpY8mTy7UPTVdQoWr03   
2246   3AVtJiu70PaFieRoxEaUfj   
2253   1p7m9H4H8s0Y7SgRm7j3ED   
2256                     None   
2269   3TPr2Em1i454nDS8UJbaFW   
2278   03jLcSeyAIVEwX2aQm5GIt   
2342   2jHpis70QAtKH3rtSvtSEu   
2384   0j9hUeyLPWahIDFbbv3D5C   
2404   4YF7DeN1nKZeN16PW9T6yJ   
2422   2dElNzSm6AR5so4wcmjavf   
2496   75JW6hFoWh8C1B34DSD87H   
2

### Analiza duration_ms
TODO

### Analiza explicit
TODO

### Analiza id_artist
TODO

In [594]:
# find tracks with null id_artist
tracks_data_frame[tracks_data_frame['id_artist'].isnull()]

id  \
26     3Z6DXqQnf2fXU9tfQqSA9j   
31     49CTqnjw0WX1iEQI0yxBVq   
80     1J1lVuoAJllePJ7lk44qKo   
95     6rPVunStILqsPHOWTv1wvB   
103    5iKjtfvlHQkmDz1wrKlVv5   
106    451AUl0iDHead9s7143WNe   
110    44estWiBus1Gw8jo1JqZHp   
120    26m9FbObHOTlXwcQH9HBH1   
141    15evaxM7psryct9lWFnYf5   
156                      None   
158    4Y4XzfNAzOqax0YcXxkP0u   
171    2XJ2otT1tK5zFlRtUMbbtF   
223    0wsxYps6jo8PgOqFyaetlS   
273    6xJrW0O0l1379Dj7imsRUf   
280    24JHtG7fAkd4scVqWFKRfe   
304    4WpRaxzrzlIhAhQ0oeAy7e   
313    7aExqfgu7jhqPIxdeQL2PE   
341    62fhOWpKKssHuLVAk8n1L3   
343    1YahNosWS6FbiOAXhlKmt7   
356    43xFiEIXdNCbvLrHnDlKSc   
365    1QQPdR1BYLoZyYe8C8XW9C   
369    2QYUqUVzRHT7r2Lmo4esW4   
378    5m3gFCiUKw7DbJYVNbwJbG   
385    762B4bOcXF7I2Y8UlKTyTy   
414                      None   
433    0erYczYtLnSRlfribXaVG1   
438    3XADapnRLrVvtNgurK8qOQ   
452                      None   
454    1ElrrSBvvhuWTvpi8aroh2   
466    7BuaeHRcyNcTUfTuq5Pr42   
482    0Ar5BVE6zI2I9NtDuqZ0uz   
492    0dH65qASmf9y6dZrjO5kY8   
503                      None   
569                      None   
574    6ukushR9Agb1l2e1V6Wtq6   
608    6V6p3kX685TlL51aeSgzpa   
633    0Vx5gIKKWhPswkRYLCNiGm   
719    5H5Sv8ARBgCMm8XBU5espL   
737    4wpr8MyRV77mffq7tv9vA8   
749    3af010A39dfMqJV708BMES   
764                      None   
804    0R5da1qbsZ8juxK3Dm8XCw   
809    4ofwffwvvnbSkrMSCKQDaC   
904    3sC3zkOUZj82KCxX6hvwPS   
928    6XzV2lWGUUKYcnVu4OA7dK   
941    5C5lsua6zIemVPDTV1QT5M   
950    37mMGMeucQCNUrA9DmhiEJ   
975    0U2RQGfkYd3IF1AT60262n   
986    4G8F3tbToecc3yriKb9aar   
998    4oJ7JAtkvIWdJbLamxYreG   
1039   1Hh2Faq4ZxuMOKUlQg6vHr   
1044   4nnMP7pBR0TJs5IeTfX2nh   
1045   0M5Rkgdoe3wWvPaXFw3jGR   
1047   5oWRC8FBcye0jXU9nfkc53   
1081   0YbfKAUSvqWj8xmvJchi9Z   
1086   09d5v2dbgYzrfXj9uxxx3r   
1105   1buASziRLNHXcTVYFWZ8EK   
1109   4iuT7B96YFBuDbyPnddDXQ   
1142   1DGuT7RBkhf2umc7YArWkM   
1151   5kcNRPGeOYWEywKj1xnnQZ   
1197   5x5fAulzj7GYJ6JbS10E8I   
1231   20806Wn1l9YseziHkB07Nu   
1269   6ASMsiaOOc5iCNP3oywskv   
1274   4BWJ5PgtovpEBRzccWOoLb   
1302   7hz1CZ9EqizK9QjAIQFqRi   
1319   3WlhsCdN4fhlqDsBEldfq8   
1328                     None   
1365   3Wz75cZysleSBZEfSgSzK5   
1375   7juQijVZXZJR4UrjGHSuIz   
1404   535wjK0TId7nPQTZpnXGgJ   
1412   2VX0VrkZyk1N2NRul0wSs0   
1484   5EcddzF6a6unGcY8Fg2NAn   
1487   1AvWd8tMGPpmXnV7eVcFFm   
1503   380FbszCi2WPjFUPfBL75n   
1512   3QcqUffnkbQGWJlfyaT3yF   
1520   1x7fkwTQE9oA1jpC6dmubF   
1526   0dQ3fWOL36Cbrpvgg1jzq3   
1532   7CH0bB9Do1XmoXxFYfuPYk   
1533   4Cj4afP6cVyJ51rW7H7lNi   
1537   3chb2A5xte2NDK3GcAytJT   
1541   7jIOye2FcVvyqSYaaTq6eR   
1545   2rPWQKW9q0mmi10LJHQ7vI   
1554   42XmWFC0UGgkfCouw6u3LX   
1556   5kBwf2GwxRHrXN55aSLstW   
1569   4fs0IGj6VbSmGVDsNE9EFQ   
1581   4xZIMRwaaBx7CZMmM6KLuh   
1599   4QXwTez2KZz4xlQoD2irHy   
1632   1NL4Hw9b0ytvmEI6yp4T1T   
1643   6KmoiS8vR3dAVrdHFzyAoO   
1648   0ahbRnj3v4RYiZ3nmnEGEH   
1650   10yp7rbTUSYI5cKcua3nc6   
1651   2QFSVtTE4odPoOImpNVyfW   
1653   4fQHH8HlNoPBxRi5vorTIf   
1682   3tDRZ3TGLZEYnQWelbW4QM   
1694   003oRl3zfRZvhttUxM9Gr5   
1705   7tqwVnj73neYmy50iW3ySK   
1777   0wJfVfp1g0Of77yJQeZ8FP   
1794   2N6hMzJBaFLQl9uYmhrEhr   
1799   21KLl43JhFp0UPZSssfm3E   
1818   5oM4tgN2A22rlHEApSElyf   
1836   3947PqotKsLhgMMx9gCOUm   
1859   3RnpduCaO9KbpZMtPOao9R   
1863   4GR4yBf2nLpUydcLLS55yX   
1880   6JlBO8RLnJHWLa2TmHnSOD   
1892   7Gp1sUO5ugcOwbiM6fpZOS   
1900   6HO2IkbKw3B7seLRuMhZl7   
1905   0TgEygavxNNlwbua53OiKE   
1929   7KmcKHOtuapNyPWq94ryAq   
1948   3JPY0GnAqbEBOadPwWEHXc   
1955   30R2nOEyimw6wG4rCFSEPc   
1959   1AOisyOwLgQ7EveVBCo0GG   
1960   1VCePFXt7vBszRThQbjhon   
1969   5RdNQsHtlAIeFeYHeH3n88   
1982   3A3n0ShUNIrNWwdzkONVAl   
1996                     None   
2000   4n4pzCKP4F8AGtCSlbJmlp   
2110   4YJcQNIIdAJL9yrtHKLCXh   
2135   4jSq795vEU1RM6AzHicKCU   
2186   4VNuoEWbaDFRb5uizQtrcS   
2187                     None   
2195   7FPJF3tt7KRzCa7HsOuA7i   
2

### Analiza release_date
TODO

### Analiza danceability
TODO

### Analiza energy
TODO

### Analiza key
TODO

### Analiza loudness
TODO

### Analiza speechiness
TODO

### Analiza acousticness
TODO

### Analiza instrumentalness
TODO

### Analiza liveness
TODO

### Analiza valence
TODO

### Analiza tempo
TODO

# Analiza użytkowników

In [595]:
users_data_frame.head(n=5)

user_id             name      city               street  \
0      101    Klara Herdzik  Szczecin   pl. Floriana 55/22   
1      102  Szymon Plebanek    Kraków   al. Kalinowa 27/40   
2      103     Szymon Zawal    Gdynia    al. Głogowa 14/10   
3      104  Andrzej Berendt    Kraków  ul. Strażacka 14/63   
4      105     Ida Karsznia    Kraków        al. Złota 549   

                      favourite_genres  premium_user   id  
0     [permanent wave, mandopop, funk]           1.0  NaN  
1                                 None           1.0 -1.0  
2      [filmi, regional mexican, folk]           1.0  NaN  
3          [tropical, new wave, latin]           1.0  NaN  
4  [soul, tropical, alternative metal]           1.0  NaN

Atrybuty:
* user_id - identyfikator użytkownika
* name - imię i nazwisko użytkownika
* city - miasto zamieszkania użytkownika
* street - ulica zamieszkania użytkownika
* favourite_genres - ulubione gatunki użytkownika
* premium_user - czy użytkownik posiada konto premium
* id - zbędny atrybut


In [596]:
users_data_frame.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   user_id           50 non-null     int64  
 1   name              50 non-null     object 
 2   city              50 non-null     object 
 3   street            50 non-null     object 
 4   favourite_genres  48 non-null     object 
 5   premium_user      49 non-null     float64
 6   id                5 non-null      float64
dtypes: float64(2), int64(1), object(4)
memory usage: 2.9+ KB


### Analiza user_id
TODO

### Analiza name
TODO

In [597]:
# find users with duplicated names
users_data_frame[users_data_frame['name'].duplicated(keep=False)]

Empty DataFrame
Columns: [user_id, name, city, street, favourite_genres, premium_user, id]
Index: []

### Analiza city
TODO

### Analiza street
TODO

### Analiza favourite_genres
TODO

In [598]:
# find users with null favourite_genres
users_data_frame[users_data_frame['favourite_genres'].isnull()]

user_id             name      city               street favourite_genres  \
1       102  Szymon Plebanek    Kraków   al. Kalinowa 27/40             None   
40      141   Apolonia Bacia  Szczecin  al. Kilińskiego 972             None   

    premium_user   id  
1            1.0 -1.0  
40           1.0  NaN

### Analiza premium_user
TODO

In [599]:
# find users with null premium_user
users_data_frame[users_data_frame['premium_user'].isnull()]

user_id         name     city                street  \
14      115  Jan Stempin  Wrocław  ulica Modrzewiowa 37   

              favourite_genres  premium_user  id  
14  [filmi, soul, turkish pop]           NaN NaN

### Analiza id
TODO
